File: Python/Recover-Binary-Search-Tree.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a tree node
    def inorder(self, root, seq):
        if root == None:
            return
        self.inorder(root.left)
        seq.append(root.val)
        self.inorder(root.right)
        return
    def recover(self, root, num1, num2):
        if root == None:
            return
        if root.val == num1:
            root.val = num2
        elif root.val == num2:
            root.val = num1
        self.recover(root.left, num1, num2)
        self.recover(root.right, num1, num2)
    def recoverTree(self, root):
        seq = []
        self.inorder(root, seq)
        index1 = 0
        index2 = 0
        for i in range(len(seq)-1):
            if seq[i] > seq[i+1]:
                index1 = i
                break
        for i in range(len(seq)-1):
            if seq[len(seq)-i-1] < seq[len(seq)-i-2]:
                index2 = len(seq) - i - 1
        index1 = seq[index1]
        index2 = seq[index2]
        self.recover(root, index1, index2)
        return

File: C++/Recover-Binary-Search-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* firstElement;
    TreeNode* secondElement;
    TreeNode* prevElement;
    void recoverTree(TreeNode* root) {
        firstElement = NULL;
        secondElement = NULL;
        prevElement = new TreeNode(INT_MIN);
        traverse(root);
        int temp = firstElement->val;
        firstElement->val = secondElement->val;
        secondElement->val = temp;
    }
    void traverse(TreeNode* root) {
        if (root == NULL) return;
        traverse(root->left);
        if (firstElement == NULL && prevElement->val >= root->val) {
            firstElement = prevElement;
        }
        if (firstElement != NULL && prevElement->val >= root->val) {
            secondElement = root;
        }
        prevElement = root;
        traverse(root->right);
    }
};